crates/sui-source-validation-service/config.toml

```diff
@@ -1,4 +1,6 @@
 [[packages]]
+source = "Repository"
+[packages.values]
```

Comment:
I'm a little confused by this syntax: `[[packages]]` above implies that `packages` is an array and you are adding an element to it, so the final structure looks like this, in JSON:

```
{ "packages": [{ "source": "Repository}] }
```

The following line, though, implies that `packages` is a dictionary, which contains a `values` field, i.e. something like this:

```
{
    "package": {
        "values": {
            "repository": "https://github.com/mystenlabs/sui",
            "branch": "mainnet-v1.3.1",
            // ... etc
        }
    }
}
```

How does the TOML parser reconcile these two shapes? They seem incompatible to me.

Reply:
I think this line on the variant helps answer?

`#[serde(tag = "source", content = "values")]`

The deserialization expects the variant to be picked by the `source` field. Then, the content, or dictionary of "values" is deserialized/populated on the corresponding variant struct (either `RepositorySource` or `DirectorySource`). I could have called it `args` or anything really, it's just the name that's used in the TOML to look for field-values to populate the given tag/variant. So this is a serde thing, that does the work of looking for the `content = "values"` bit in the config.

So, e.g., in the test I added to the same TOML config

```toml
    [[packages]]
    source = "Directory"
    [packages.values]
    paths = [ "home/user/some/package" ]
```

for the directory case. And the test output is the serialized Rust representation (no `values` field occurs anywhere).

What the serialized JSON of this looks like I'm not sure (i.e., whether it adds a `values` dictionary). Right now we have no need to serialize so I didn't really dig into that.

Reply:
FWIW this serde way is one way I found how to deserialize to Rust variants when I went looking for it. I've seen other languages deal with this a bit differently for JSON/TOML/etc, but I figured there had to be something out there for Rust and this seemed quite fine to me.

Reply:
I ran the example through [this converter](https://pseitz.github.io/toml-to-json-online-converter/), and it helps clarify my confusion:  This syntax actually adds fields to the dictionary at the end of the packages array, rather than adding fields to `packages` itself, TIL!

Reply:
(My confusion was purely about the TOML syntax by the way, not how it was deserialized into Rust, and converting it to JSON was just my way of parsing out the structure into a more explicit form, rather than because I expected us to need to do that in this implementation).

Reply:
ah, misread the question, sorry about that!